### PR TITLE
Triton MLA perf fixes

### PR DESCRIPTION
**Pull request by koush**
## Purpose

Triton MLA on sm120 performance degrades on batch size 1 as context length increases.

This perf issue has been bugging me for a while as it made deepseek and Kimi k2 unusable, and when Kimi k2.5 was released I finally got around to digging into it. I'm familiar with w/ cuda but this is my first foray into triton. 

The primary issues are suboptimal kv splitting during low batch count resulting in underutilized SM and unnecessary load of Q vector, since c_kv contains both.

## Test Plan

Test Kimi 2.5, Deepseek v2, Qwen 235B.

## Test Result

Models load and generate correctly, Kimi 2.5 shows notable improvement as seen below.

Testing with -dcp 8 and 80k context

| Concurrent Requests | OLD Peak Throughput (tok/s) | NEW Peak Throughput (tok/s) | Peak Throughput Increase (%) |
|---------------------|----------------------------|----------------------------|------------------------------|
| 1 (128 tokens)      | 54.00                      | 55.00                      | 1.85%                        |
| 8 (128 tokens)      | 272.00                     | 272.00                     | 0.00%                        |
| 1 (80K tokens)      | 34.00                      | 52.00                      | 52.94%                       |
| 8 (80K tokens)      | 176.00                     | 184.00                     | 4.55%                        |

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [X] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [X] The test plan, such as providing test command.
- [X] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/v1/attention/backends/mla/triton_mla.py [hunk 1]
```diff
@@ -134,8 +146,14 @@ def forward_mqa(
         )
         lse = torch.zeros(B, q_num_heads, dtype=q.dtype, device=q.device)
 
-        # For batch invariance, use only 1 split to ensure deterministic reduction
-        num_kv_splits = 1 if vllm_is_batch_invariant() else 4
+        if vllm_is_batch_invariant():
+            num_kv_splits = 1
+        else:
+            properties = torch.cuda.get_device_properties(q.device)
```
Comment (koush) [inline vllm/v1/attention/backends/mla/triton_mla.py]: This is the primary fix. Under utilization of the available sm. Changing num_kv_splits to be dynamic like the other implementations got the token generation at 80k context from 5t/s to 20t/s.

File: vllm/v1/attention/ops/triton_decode_attention.py [hunk 2]
```diff
@@ -295,7 +295,12 @@ def _fwd_grouped_kernel_stage1(
     cur_batch_req_idx = cur_batch
 
     offs_q = cur_batch * stride_qbs + cur_head[:, None] * stride_qh + offs_d[None, :]
-    q = tl.load(Q + offs_q, mask=(mask_h[:, None]) & (mask_d[None, :]), other=0.0)
+    q = tl.load(
```
Comment (koush) [inline vllm/v1/attention/ops/triton_decode_attention.py:314]: Explicitly providing cache hints to the load operations t/s a by 2-3t/s.

File: vllm/v1/attention/ops/triton_decode_attention.py [hunk 3]
```diff
@@ -316,38 +324,58 @@ def _fwd_grouped_kernel_stage1(
...
-                kv_loc[None, :] * stride_buf_kbs
+                kv_loc[:, None] * stride_buf_kbs
                 + cur_kv_head * stride_buf_kh
-                + offs_d[:, None]
+                + offs_d[None, :]
             )
             k = tl.load(
                 K_Buffer + offs_buf_k,
-                mask=(offs_n[None, :] < split_kv_end) & (mask_d[:, None]),
+                mask=(offs_n[:, None] < split_kv_end) & (mask_d[None, :]),
```
Comment (koush) [inline vllm/v1/attention/ops/triton_decode_attention.py]: Examination of the PTX revealed non contiguous 8 byte read access patterns. Same as below. Load contiguously then transpose. Resulted In significantly smaller ptx size (but almost nearly no gain because I suspect due to l1/l2 cache hits). PTX size went from 140k to 110k with less operations.

File: vllm/v1/attention/ops/triton_decode_attention.py [hunk 4]
```diff
@@ -316,38 +324,58 @@ def _fwd_grouped_kernel_stage1(
     acc = tl.zeros([BLOCK_H, BLOCK_DV], dtype=tl.float32)
 
     if split_kv_end > split_kv_start:
-        for start_n in range(split_kv_start, split_kv_end, BLOCK_N):
+        for start_n in tl.range(split_kv_start, split_kv_end, BLOCK_N):
```
Comment (koush) [inline vllm/v1/attention/ops/triton_decode_attention.py:350]: This doesn't have any effect that I could see, but tl.range is the idiomatic way to loop in triton and provides compiler hints.

File: vllm/v1/attention/ops/triton_decode_attention.py [hunk 5]
```diff
@@ -316,38 +324,58 @@ def _fwd_grouped_kernel_stage1(
...
             )
             k = tl.load(
                 K_Buffer + offs_buf_k,
-                mask=(offs_n[None, :] < split_kv_end) & (mask_d[:, None]),
+                mask=(offs_n[:, None] < split_kv_end) & (mask_d[None, :]),
                 other=0.0,
+                cache_modifier=".cg",
             )
+
+            offs_buf_v = (
```
Comment (koush) [inline vllm/v1/attention/ops/triton_decode_attention.py]: Triton seemed to have a difficult time overlapping load and compute here, and this explicit front loading helped by another 1-2t/s.

File: vllm/v1/attention/backends/mla/triton_mla.py [hunk 6]
```diff
@@ -26,6 +28,12 @@
 logger = init_logger(__name__)
 
 
+class TritonMLAMetadataBuilder(MLACommonMetadataBuilder[MLACommonMetadata]):
+    _cudagraph_support: ClassVar[AttentionCGSupport] = (
```
Comment (koush) [inline vllm/v1/attention/backends/mla/triton_mla.py]: I'm not sure why cuda graph support was omitted, seems to be an oversight. Added here.

**Comment by koush** [comment]
benchmarks on Kimi k2.5 

Summary

 | Context | OLD (tok/s) | NEW (tok/s) | Improvement |
|---------|-------------|-------------|-------------|
| Short (100 in / 400 out) | 69.14 | 72.37 | +4.7% |
| Long (80K in / 400 out) | 5.61 | 9.61 | +71.3% |

```
OLD

============ Serving Benchmark Result ============
Successful requests:                     2         
Failed requests:                         0         
Maximum request concurrency:             1         
Benchmark duration (s):                  11.57     
Total input tokens:                      200       
Total generated tokens:                  800       
Request throughput (req/s):              0.17      
Output token throughput (tok/s):         69.14     
Peak output token throughput (tok/s):    72.00     
Peak concurrent requests:                2.00      
Total token throughput (tok/s):          86.42     
---------------Time to First Token----------------
Mean TTFT (ms):                          76.94     
Median TTFT (ms):                        76.94     
P99 TTFT (ms):                           81.87     
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          14.31     
Median TPOT (ms):                        14.31     
P99 TPOT (ms):                           14.45     
---------------Inter-token Latency----------------
Mean ITL (ms):                           14.31     
Median ITL (ms):                         14.15     
P99 ITL (ms):                            20.66     
==================================================



============ Serving Benchmark Result ============
Successful requests:                     2         
Failed requests:                         0         
Maximum request concurrency:             1         
Benchmark duration (s):                  142.69    
Total input tokens:                      160000    
Total generated tokens:                  800       
Request throughput (req/s):              0.01      
Output token throughput (tok/s):         5.61      
Peak output token throughput (tok/s):    9.00      
Peak concurrent requests:                2.00      
Total token throughput (tok/s):          1126.90   
---------------Time to First Token----------------
Mean TTFT (ms):                          25966.21  
Median TTFT (ms):                        25966.21  
P99 TTFT (ms):                           25980.01  
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          113.73    
Median TPOT (ms):                        113.73    
P99 TPOT (ms):                           113.86    
---------------Inter-token Latency----------------
Mean ITL (ms):                           113.73    
Median ITL (ms):                         113.74    
P99 ITL (ms):                            114.33    
==================================================

NEW

============ Serving Benchmark Result ============
Successful requests:                     2         
Failed requests:                         0         
Maximum request concurrency:             1         
Benchmark duration (s):                  11.06     
Total input tokens:                      200       
Total generated tokens:                  800       
Request throughput (req/s):              0.18      
Output token throughput (tok/s):         72.37     
Peak output token throughput (tok/s):    74.00     
Peak concurrent requests:                2.00      
Total token throughput (tok/s):          90.46     
---------------Time to First Token----------------
Mean TTFT (ms):                          74.98     
Median TTFT (ms):                        74.98     
P99 TTFT (ms):                           78.82     
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          13.66     
Median TPOT (ms):                        13.66     
P99 TPOT (ms):                           13.68     
---------------Inter-token Latency----------------
Mean ITL (ms):                           13.66     
Median ITL (ms):                         13.66     
P99 ITL (ms):                            14.08     
==================================================

============ Serving Benchmark Result ============
Successful requests:                     2         
Failed requests:                         0         
Maximum request concurrency:             1         
Benchmark duration (s):                  83.21     
Total input tokens:                      160000    
Total generated tokens:                  800       
Request throughput (req/s):              0.02      
Output token throughput (tok/s):         9.61      
Peak output token throughput (tok/s):    26.00     
Peak concurrent requests:                2.00      
Total token throughput (tok/s):          1932.46   
---------------Time to First Token----------------
Mean TTFT (ms):                          26040.45  
Median TTFT (ms):                        26040.45  
P99 TTFT (ms):                           26041.25  
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          39.01     
Median TPOT (ms):                        39.01     
P99 TPOT (ms):                           39.03     
---------------Inter-token Latency----------------
Mean ITL (ms):                           39.01     
Median ITL (ms):                         39.00     
P99 ITL (ms):                            39.52     
==================================================


```

File: vllm/v1/attention/backends/mla/triton_mla.py [hunk 7]
```diff
@@ -134,8 +146,13 @@ def forward_mqa(
         )
         lse = torch.zeros(B, q_num_heads, dtype=q.dtype, device=q.device)
 
-        # For batch invariance, use only 1 split to ensure deterministic reduction
-        num_kv_splits = 1 if vllm_is_batch_invariant() else 4
+        if vllm_is_batch_invariant():
+            num_kv_splits = 1
+        else:
+            properties = torch.cuda.get_device_properties(q.device)
+            num_sms = properties.multi_processor_count
```
Comment (mgoin) [inline vllm/v1/attention/backends/mla/triton_mla.py]: This should be accessed and saved outside of forward path, see cutlass_mla or flashmla for reference

File: vllm/v1/attention/backends/mla/triton_mla.py [hunk 8]
```diff
@@ -98,6 +110,11 @@ def __init__(
                 "TritonMLA V1 with FP8 KV cache not yet supported"
             )
 
+        # Pre-compute sm_count to avoid recomputing it. Use device 0 as a proxy
+        # (assumes all devices are similar)
+        properties = torch.cuda.get_device_properties(torch.device("cuda:0"))
```
Comment (tjtanaa) [inline vllm/v1/attention/backends/mla/triton_mla.py]: Let's use the helper function `get_cu_count` from `vllm/vllm/utils/platform_utils.py`

https://github.com/vllm-project/vllm/blob/1b60b45d0d745e0c0be65994507a702a19fd4761/vllm/utils/platform_utils.py#L27

File: vllm/v1/attention/backends/mla/triton_mla.py (hunk 6 above)
Comment (tjtanaa) [reply vllm/v1/attention/backends/mla/triton_mla.py]: @koush Another thing, iirc DCP and PCP is not graph compatible, as stated in the code. Can you find out if this changes the default behaviour? Do we need to explicitly state `--compilation-config '{"cudagraph_mode": "PIECEWISE"}' ` after this changes?

`vllm serve Qwen/Qwen3-235B-A22B -tp8 -dcp 2`

image

Comment (koush) [reply vllm/v1/attention/backends/mla/triton_mla.py]: I'm unaware of these arguments, which docs is this in? I didnt find it in the vllm repo, docs, or in the kimi model deployment guide.

https://huggingface.co/moonshotai/Kimi-K2.5/blob/main/docs/deploy_guidance.md

File: vllm/v1/attention/backends/mla/triton_mla.py (hunk 8 above)
Comment (koush) [reply vllm/v1/attention/backends/mla/triton_mla.py]: fixed

**Comment by koush** [comment]
Testing with -dcp 8 and 80k context

| Concurrent Requests | OLD Output Throughput (tok/s) | NEW Output Throughput (tok/s) | OLD Peak Throughput (tok/s) | NEW Peak Throughput (tok/s) | Output Throughput Increase (%) | Peak Throughput Increase (%) |
|--------------------:|------------------------------:|------------------------------:|----------------------------:|----------------------------:|-------------------------------:|-----------------------------:|
| 1                   | 27.32                         | 36.66                         | 34.00                       | 49.00                       | **+34.2%**                     | **+44.1%**                   |
| 8                   | 43.47                         | 50.60                         | 60.00                       | 72.00                       | **+16.4%**                     | **+20.0%**                   |

File: vllm/v1/attention/backends/mla/triton_mla.py [hunk 9]
```diff
@@ -134,8 +138,27 @@ def forward_mqa(
...
+
+            # use power of 2 to avoid excessive kernel instantiations
+            ideal_splits = triton.next_power_of_2(ideal_splits)
+
+            # Calculate SM-based maximum splits with occupancy multiplier
+            # 2-4x allows multiple blocks per SM for latency hiding
+            # hardware dependent
+            occupancy_multiplier = 2
+            max_splits = self._sm_count * occupancy_multiplier
+            num_kv_splits = min(ideal_splits, max_splits)
```
Comment (koush) [inline vllm/v1/attention/backends/mla/triton_mla.py:172]: This is the primary fix of this change. num_kv_splits of 1/4 is not enough to saturate SMs at long context. It wholly relied on batch size for parallelism.

**Review by mgoin — APPROVED** [review]
Okay nice work, this looks reasonable to me now. Can you run an accuracy evaluation? Just a simple gsm8k is good. Another point would be to test GLM-4.7-Flash since it has a unique head size compared to deepseek/kimi MLA

**Review by LucasWilkinson — APPROVED** [review]
LGTM

**Comment by koush** [comment]
> gsm8k

I assume you mean this?

This? https://docs.vllm.ai/projects/ascend/en/latest/developer_guide/evaluation/using_lm_eval.html

**Review by MatthewBonanni — APPROVED** [review]
LGTM
